### DISCUSSION TITLE
OpenAPI document has a different PORT

### DISCUSSION THREAD

**Discussion by megashrieks** [microsoft/aspire#6989] [General]
I'm on dotnet 9.0, and somehow `AddOpenApi` creates an openapi spec that has a different port specified in the document

and each time the application starts the port is assigned randomly, I'm guessing that this might somehow be related to the proxy created by aspire?

I use scalar, and this is annoying since everytime I need to call an API after restarting my application, I need to go to
Aspire Dashboard > My application scalar end point > call the api

since upon restart the port changes everytime I cant just keep the scalar page loaded

any idea on how I can disable this? or is this occuring from Aspire at all? 

here is my launchsettings.json
```
{
  "$schema": "http://json.schemastore.org/launchsettings.json",
  "profiles": {
    "http": {
      "commandName": "Project",
      "dotnetRunMessages": true,
      "launchBrowser": true,
      "launchUrl": "scalar/v1",
      "applicationUrl": "http://localhost:8080",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}
```
openapi generated
```
{
  "openapi": "3.0.1",
  "info": {
    "title": "Application.WebApi | v1",
    "version": "1.0.0"
  },
  "servers": [
    {
      "url": "http://localhost:33815"
    }
  ],
  "paths": {
```

Currently I've hardcoded the port in Openapi document transformer, but I would prefer it if there was a better solution to this

**Comment by davidfowl** [comment]
See https://github.com/dotnet/aspnetcore/issues/57332. Yes it because of the proxy. The open api support in asp.net core should handle this better.

**Reply by megashrieks** [reply]
okay thank you

**Comment by TheBenda** [comment]
As none of the suggestions from https://github.com/dotnet/aspnetcore/issues/57332 worked for me I looked into the cause myself and found a very convenient workaround by reading the actual url from the env-vars and setting it manually:

```
app.MapScalarApiReference("/api-reference", options => 
{
    options.WithTitle("Some REST API")
        .WithTheme(ScalarTheme.Moon)
        .WithDefaultHttpClient(ScalarTarget.CSharp, ScalarClient.HttpClient);
    
    List<ScalarServer> servers = []; 
    var proxyUrl = Environment.GetEnvironmentVariable("ASPNETCORE_URLS"); 
    if (proxyUrl is not null) { 
        servers.Add(new ScalarServer(proxyUrl)); 
    } 

    options.Servers = servers;
});
```

Optionally you can set the servers in the `AddOpenApi(...)` definition as well.

**Comment by davidfowl** [comment]
cc @captainsafia

**Comment by captainsafia** [comment]
This has been fixed in .NET 9.0.4 and .NET 10 Preview 3.

Note: that for use with Aspire you'll also have to configure the [forwarded headers middleware](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-9.0#forwarded-headers-middleware-order) to get the desired behavior.

**Reply by svha1977** [reply]
Any options for Azure functions isolated .net 9.0. Has the same issue.
